pytato/target/pycuda/cudagraph.py

```diff
@@ -0,0 +1,83 @@
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import ast
+
+from typing import Union, Optional, Mapping
+
+from pytato.array import Array, DictOfNamedArrays
+from pytato.target.pycuda import CUDAGraphTarget, BoundCUDAGraphProgram
+from pytato.target.pycuda.numpy_like import generate_numpy_like
+
+__doc__ = """
+.. autofunction:: generate_cudagraph
+"""
+
+def generate_cudagraph(expr: Union[Array, Mapping[str, Array], DictOfNamedArrays],
+                 *,
+                 target: Optional[CUDAGraphTarget] = None,
+                 function_name: str = "_pt_kernel",
+                 show_code: bool = False,
+                 colorize_show_code: bool = True,
+                 ) -> BoundCUDAGraphProgram:
+    """
+    Returns a :class:`pytato.target.python.BoundCUDAGraphProgram` for the array
+    expressions in *expr*.
+    :arg function: Name of the entrypoint function in the generated code.
+    :arg show_code: If *True*, the generated code is printed to ``stdout``.
+    """
+    if target is None:
+        target = CUDAGraphTarget()
+
+    extra_preambles = []
+    extra_preambles.append(ast.ImportFrom(module="pycuda",
+                                              names=[ast.alias(
+                                                  "elementwise",
+                                                  asname="elementwise")],
+                                              level=0))
+    
+    extra_preambles.append(ast.ImportFrom(module="pycuda",
+                                            names=[ast.alias(
+                                                "reduction",
+                                                asname="reduction")],
+                                            level=0))
+    
+    extra_preambles.append(ast.ImportFrom(module="pycuda",
+                                            names=[ast.alias(
+                                                "gpuarray",
+                                                asname="")],
+                                            level=0))
+
+    extra_preambles.append(ast.ImportFrom(module="pycuda",
+                                              names=[ast.alias(
+                                                  "autoinit",
+                                                  asname="")],
+                                              level=0))
+
+
```

Comment:
These lines should end up in whatever `generate_numpy_like` gets renamed to.

Reply:
`_pt` is a protected namespace in pytato. So any non-user facing variable should be named that i.e. `_pt_cuda_elemwise`, `_pt_cuda_elemwise`, etc.